Implement Zicbom, Zicboz (cbo.flush, cbo.inval, cbo.zero)

Timmmm opened the pull request:
Note that Zicbop (prefetch hints) does not need to be implemented because all it does is label some existing base instructions as prefetch hints.

~~Also I have not wired up the enable flags to the emulators because it is rather tedious (and will hopefully be replaced by riscv-config at some point).~~

arichardson commented on model/riscv_insts_zicboz.sail:
```diff
@@ -0,0 +1,66 @@
+    let cache_block_size_exp = plat_cache_block_size_exp();
+    let cache_block_size = 2 ^ cache_block_size_exp;
+
+    // Offset to the beginning of the cache block.
+    let offset = (rs1_val & ~(zero_extend(ones(cache_block_size_exp)))) - rs1_val;
+
+    match ext_data_get_addr(rs1, offset, Write(Data), cache_block_size) {
```
```suggestion
    // Offset from the beginning of the cache block.
    let offset = rs1_val - (rs1_val & ~(zero_extend(ones(cache_block_size_exp))));

    match ext_data_get_addr(rs1, -offset, Write(Data), cache_block_size) {
```
Not sure if this is any better, but I find it easier to understand if we subtract the offset in the call to ext_data_get_addr.

Timmmm replied:
Actually... there isn't a unary `-` operator for bit vectors so you have to do `zeros() - offset` which is just a bit weird, so I changed it back.

arichardson commented on model/riscv_insts_zicbom.sail:
```diff
@@ -0,0 +1,107 @@
+  // Offset to the beginning of the cache block.
+  let offset = (rs1_val & ~(zero_extend(ones(cache_block_size_exp)))) - rs1_val;
+
+  match ext_data_get_addr(rs1, offset, Read(Data), cache_block_size) {
```
This probably needs a new hook for CHERI to allow "any byte in the cacheblock" rather than all bytes being covered.

E.g. extra_data_check_cmo_addr(rs1, Read(Data)) that just forwards to `ext_data_get_addr()` in the base RISC-V model?

Timmmm replied:
Good point, though I am not 100% sure on the intended CHERI behaviour - see https://github.com/riscv/riscv-cheri/issues/213

I added a `// TODO` at least so it hopefully doesn't get forgotten.

bacam commented:
> Note that Zicbop (prefetch hints) does not need to be implemented because all it does is label some existing base instructions as prefetch hints.

Shouldn't there be encoding and assembly information at least?  It might also be useful to calculate the address and pass it to a built-in function so that it could be logged (e.g., to check that some code's prefetch instruction refers to the correct address), along with a comment noting that no exceptions are taken if something goes wrong.

Timmmm commented on c_emulator/riscv_sim.c:
```diff
@@ -405,6 +408,18 @@ static int process_args(int argc, char **argv)
       fprintf(stderr, "using %s for trace output.\n", trace_log_path);
+    case OPT_CACHE_BLOCK_SIZE_EXP:
+      block_size_exp = atol(optarg);
+      if (block_size_exp <= 12) {
```
Yeah I agree that would probably be easier, though... my OCaml is not really that good... I'll change it if you can give me a hint how to do it in OCaml! :-D

Timmmm replied:
Ok I have updated this now as you suggested.

Timmmm commented:
> Shouldn't there be encoding and assembly information at least? It might also be useful to calculate the address and pass it to a built-in function so that it could be logged (e.g., to check that some code's prefetch instruction refers to the correct address), along with a comment noting that no exceptions are taken if something goes wrong.

That's a good point... I think it can be a separate PR though if anyone wants to tackle it.

animesh211 commented:
HI, 
Just wanted to check, Is this the main PR for cbo.zero support in sail? and Can anyone please let me know, if there's an ETA to close this.

Timmmm commented:
Yes it is - it's based on the older one (#137) with some fixes. As far as I am concerned it is ready to merge; it's just waiting on review.

No ETA I'm afraid - not many people are actively reviewing PRs and unfortunately nobody is paid to do it currently. (IMO RVI should hire someone to do this but that's an ongoing discussion.)

billmcspadden-riscv commented:
@Tim Hutt ***@***.***> ,  how did you test your
implementation?

Bill Mc.


On Tue, Jul 16, 2024 at 7:45 AM Bill McSpadden ***@***.***> wrote:

> I'll review it.
>
> However,  I see that there are some conflicts that need to be resolved.
>
> Bill Mc.
>
>
> On Tue, Jul 16, 2024 at 6:18 AM Tim Hutt ***@***.***> wrote:
>
>> Yes it is - it's based on the older one (#137
>> <https://github.com/riscv/sail-riscv/pull/137>) with some fixes. As far
>> as I am concerned it is ready to merge; it's just waiting on review.
>>
>> No ETA I'm afraid - not many people are actively reviewing PRs and
>> unfortunately nobody is paid to do it currently. (IMO RVI should hire
>> someone to do this but that's an ongoing discussion.)
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/riscv/sail-riscv/pull/455#issuecomment-2230642606>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AXROLOETFGKT3KKCYQHYNCLZMT6PZAVCNFSM6AAAAABGTDUZUKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMZQGY2DENRQGY>
>> .
>> You are receiving this because you are subscribed to this thread.Message
>> ID: ***@***.***>
>>
>
>
> --
> Bill McSpadden
> Formal Verification Engineer
> RISC-V International
> mobile:  503-807-9309
>


-- 
Bill McSpadden
Formal Verification Engineer
RISC-V International
mobile:  503-807-9309

Timmmm commented:
Thanks, I've rebased and fixed the conflicts.

It's been tested by running against our A730 chip with the following tests:

* `cbo.zero-01.S` from `riscv-arch-test`
* `zero.S` from `riscv-tests`
* A very basic directed test we wrote that does some stores and loads combined with the CBO instructions and CSR settings.
* STING - it's not really clear exactly what they test (the descriptions are unhelpfully short like "Test for cmo instruction - CBO.FLUSH") but they're generally pretty good.

The first two only test `cbo.zero`; I'm not sure why they didn't add tests for the other instructions.

jrtc27 commented on model/riscv_sys_regs.sail:
```diff
@@ -108,6 +108,14 @@ val sys_enable_vext = {c: "sys_enable_vext", ocaml: "Platform.enable_vext", _: "
 
+// CBO extensions. Zicbop cannot be enabled/disabled because it has no effect
+// at all on this model.
+// TODO: Add command line flags.
```
IIRC one of these conflicts with the current CHERI-RISC-V encoding for LC (and the community's historic unofficial consensus for RV128's LQ, which is where that came from), so sail-cheri-riscv needs this to be disableable in sail-riscv without code modification to riscv_sys_regs.sail.

Timmmm replied:
Ah yeah, you're right. `cbo.zero` is

```
mapping clause encdec = RISCV_ZICBOZ(rs1)                if haveZicboz()
  <-> 0b000000000100 @ rs1 @ 0b010 @ 0b00000 @ 0b0001111 if haveZicboz()
```

And that used to conflict with `lc`, which was this (I think `cd != zreg` was our temporary workaround):

```
mapping clause encdec = LoadCapImm(cd, rs1, offset) if haveXcheri() & sizeof(xlen) == 64 & cd != zreg  /* lc */
   <-> offset @ rs1 @ 0b010 @ cd @ 0b0001111 if haveXcheri() & sizeof(xlen) == 64 & cd != zreg
```

However that was [changed](https://riscv.github.io/riscv-cheri/#LC) so it's now this:

```
mapping clause encdec = LoadCapImm(cd, rs1, offset) if haveXcheri() & sizeof(xlen) == 64  /* lc */
  <-> offset @ rs1 @ 0b100 @ cd @ 0b0001111         if haveXcheri() & sizeof(xlen) == 64
```

jrtc27 replied:
Yes but it's going to take a while to shift everything over to that...

Timmmm replied:
Ok I'll add command line flags.

Timmmm replied:
Done

jrtc27 commented on model/riscv_insts_zicbom.sail:
```diff
@@ -0,0 +1,110 @@
+
+// Cache Block Operations - Management
+//
+// Note that the CBOP prefetch instructions are existing base instructions that
```
Why would anyone expect Zicbop instructions to be in Zicbom or Zicboz files?

jrtc27 replied:
(and this is false anyway, you still want correct disassembly, and some stub Sail code)

Timmmm replied:
Good point, I removed that comment.

jrtc27 commented on model/riscv_insts_zicbom.sail:
```diff
@@ -0,0 +1,110 @@
+// (since this model has no caches to prefetch into and the encodings already
+// exist).
+
+function cbo_clean_flush_enabled(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBCFE()[0], senvcfg.CBCFE()[0])
```
Old bitfield syntax

Timmmm replied:
Fixed

jrtc27 commented on model/riscv_insts_zicbom.sail:
```diff
@@ -0,0 +1,110 @@
+  let cache_block_size = 2 ^ cache_block_size_exp;
+
+  // Offset from the beginning of the cache block.
+  let offset = (rs1_val & ~(zero_extend(ones(cache_block_size_exp)))) - rs1_val;
```
This is negative when not aligned?

jrtc27 replied:
Ok, I see that's intentional to compensate in the ext_data_get_addr call to offset back to the start of the cache block. Which means the *comment* here is wrong, it's the offset *to* the beginning of the cache block, not from.

Timmmm replied:
Yeah it's quite confusing. I expanded & fixed the comment.

jrtc27 commented on model/riscv_insts_zicbom.sail:
```diff
@@ -0,0 +1,110 @@
+          let exc_read = phys_access_check(Read(Data), cur_privilege, paddr, cache_block_size);
+          let exc_write = phys_access_check(Write(Data), cur_privilege, paddr, cache_block_size);
+          match (exc_read, exc_write) {
+            (Some(e0), Some(e1)) => Some(e1),
```
Where is it specified that one should prefer the write exception?

jrtc27 replied:
Also, if one fails but not the other, this treats that as a success? That seems odd.

jrtc27 replied:
I guess that's from this:
> If access to the cache block is not permitted, a cache-block management instruction raises a store page fault or store guest-page fault exception if address translation does not permit any access or raises a store access fault exception otherwise.

Might be worth explaining that.

Timmmm replied:
Yep, comment added.

jrtc27 commented on model/riscv_insts_zicbom.sail:
```diff
@@ -0,0 +1,110 @@
+}
+
+function clause execute(RISCV_ZICBOM(cbop, rs1)) = {
+  if (cbop == CBO_CLEAN | cbop == CBO_FLUSH) & cbo_clean_flush_enabled(cur_privilege) then {
```
match?

Timmmm replied:
Yeah I wasn't sure about this. It means you have to duplicate a branch because (as far as I know) Sail doesn't support OR patterns, (`match foo { A | B => ...`). Unless I'm wrong about that?

Anyway I updated it to `match` with a duplicate branch. Maybe slightly better.

jrtc27 commented on model/riscv_platform.sail:
```diff
@@ -26,6 +26,9 @@ val elf_entry = {
   c: "elf_entry"
 } : unit -> int
 
+// Cache block size is 2^cache_block_size_exp. Max 4096.
```
Perhaps worth documenting why this is the implementation limit (minimum page size)? I don't see it required in the spec.

Timmmm replied:
Yeah it comes from `max_mem_access`, which I set to 4096 (it was 16 previously). Reading the spec it doesn't really mention the behaviour for cache blocks larger than a page (presumably because they thought that was not a sane configuration and didn't think about it), but I think it would bring up awkward questions if you tried to do that, so I just limited at 1 page.

I added a comment to hopefully explain that.

Timmmm commented:
Thanks for the review @jrtc27 !

animesh211 commented:
HI, It'll be great if this can be reviewed and closed.

aprnath commented:
Adding a gentle nudge to animesh211's request for this to be merged and closed.

arichardson commented on c_emulator/riscv_sim.c:
```diff
@@ -414,7 +452,6 @@ static int process_args(int argc, char **argv)
     case OPT_TRACE_OUTPUT:
       trace_log_path = optarg;
       fprintf(stderr, "using %s for trace output.\n", trace_log_path);
-      break;
```
why is this line gone?

Timmmm replied:
Good spot - probably a rebase mistake. I'll fix it.

Timmmm replied:
Fixed.

arichardson commented on c_emulator/riscv_sim.c:
```diff
@@ -234,6 +240,17 @@ static void read_dtb(const char *path)
+static int ilog2(uint64_t x)
+{
+  for (unsigned i = 0; i < sizeof(x) * 8; ++i) {
+    if ((1ull << i) == x) {
```
```suggestion
    if (x == (UINT64_C(1) << i)) {
```

arichardson commented on model/riscv_insts_zicboz.sail:
```diff
@@ -0,0 +1,64 @@
+    // is aligned to the cache block, or negative if rs1 is misaligned.
+    let offset = (rs1_val & ~(zero_extend(ones(cache_block_size_exp)))) - rs1_val;
+
+    // TODO: This is not correct since CHERI treats CBO operations
+    // differently to normal accesses with respect to bounds checks.
+    match ext_data_get_addr(rs1, offset, Write(Data), cache_block_size) {
```
```suggestion
    match ext_data_get_addr(rs1, offset, Write(Data), cache_block_size) {
```
CBOZ should be correct for CHERI since we check all bytes?

Timmmm commented:
I rebased this and updated it to use the new `extensionEnabled()` function. Arguably this:

```
function clause extensionEnabled(Ext_Zicbom) = sys_enable_zicbom()
```

should be this

```
function clause extensionEnabled(Ext_Zicbom) = sys_enable_zicbom() & cbo_zero_enabled(cur_privilege)
```

however I left it as it was just so we can get this in.

@billmcspadden-riscv any objection to merging this? I think it has been ready with no objections for a long time.

billmcspadden-riscv reviewed (approved):
LGTM

Timmmm commented:
Thanks!
